src/kademlia/dht_storage.cpp

```diff
@@ -0,0 +1,568 @@
+				++f->num_announcers;
+			}
+		}
+
+		// return true of the first argument is a better canidate for removal, i.e.
+		// less important to keep
+		struct immutable_item_comparator
+		{
+			immutable_item_comparator(node_id const& our_id) : m_our_id(our_id) {}
+			immutable_item_comparator(immutable_item_comparator const& c)
+				: m_our_id(c.m_our_id) {}
+
+			bool operator() (std::pair<node_id, dht_immutable_item> const& lhs
+				, std::pair<node_id, dht_immutable_item> const& rhs) const
+			{
+				int l_distance = distance_exp(lhs.first, m_our_id);
+				int r_distance = distance_exp(rhs.first, m_our_id);
+
+				// this is a score taking the popularity (number of announcers) and the
+				// fit, in terms of distance from ideal storing node, into account.
+				// each additional 5 announcers is worth one extra bit in the distance.
+				// that is, an item with 10 announcers is allowed to be twice as far
+				// from another item with 5 announcers, from our node ID. Twice as far
+				// because it gets one more bit.
+				return lhs.second.num_announcers / 5 - l_distance < rhs.second.num_announcers / 5 - r_distance;
+			}
+
+		private:
+
+			// explicitly disallow assignment, to silence msvc warning
+			immutable_item_comparator& operator=(immutable_item_comparator const&);
+
+			node_id const& m_our_id;
+		};
+	}
+
+	class dht_default_storage : public dht_storage_interface, boost::noncopyable
+	{
+	typedef std::map<node_id, torrent_entry> table_t;
+	typedef std::map<node_id, dht_immutable_item> dht_immutable_table_t;
+	typedef std::map<node_id, dht_mutable_item> dht_mutable_table_t;
+
+	public:
+
+		dht_default_storage(sha1_hash const& id, dht_settings const& settings
+			, counters& cnt);
+
+		~dht_default_storage();
+
+#ifndef TORRENT_NO_DEPRECATE
+		size_t num_torrents() const;
+		size_t num_peers() const;
+#endif
+
+		bool lookup_peers(sha1_hash const& info_hash
+			, bool noseed, bool scrape
+			, entry& peers) const;
+		void announce_peer(sha1_hash const& info_hash
+			, tcp::endpoint const& endp
+			, std::string const& name, bool seed);
+
+		bool get_immutable_item(sha1_hash const& target
+			, entry& item) const;
+		void put_immutable_item(sha1_hash const& target
+			, char const* buf, int size
+			, address const& addr);
+
+		bool get_mutable_item_seq(sha1_hash const& target
+			, boost::int64_t& seq) const;
+		bool get_mutable_item(sha1_hash const& target
+			, boost::int64_t seq, bool force_fill
+			, entry& item) const;
+		void put_mutable_item(sha1_hash const& target
+			, char const* buf, int size
+			, char const* sig
+			, boost::int64_t seq
+			, char const* pk
+			, char const* salt, int salt_size
+			, address const& addr);
+
+		void tick();
+
+	private:
+		sha1_hash m_id;
+		dht_settings const& m_settings;
+		counters& m_counters;
+
+		table_t m_map;
+		dht_immutable_table_t m_immutable_table;
+		dht_mutable_table_t m_mutable_table;
+	};
+
+	dht_default_storage::dht_default_storage(sha1_hash const& id
```

Comment:
for classes that are private in a translation unit, I normally put all the implementations inline (to not have to write the signatures twice).